internal/hcp/internal/controllers/link/controller.go

```diff
@@ -15,17 +19,69 @@ import (
 	pbhcp "github.com/hashicorp/consul/proto-public/pbhcp/v2"
 )
 
-func LinkController(resourceApisEnabled bool, hcpAllowV2ResourceApis bool) *controller.Controller {
+type HCPClientFn func(link *pbhcp.Link) (hcpclient.Client, error)
+
+func NewHCPClient(link *pbhcp.Link) (hcpclient.Client, error) {
+	hcpClient, err := hcpclient.NewClient(config.CloudConfig{
+		ResourceID:   link.ResourceId,
+		ClientID:     link.ClientId,
+		ClientSecret: link.ClientSecret,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return hcpClient, nil
+}
+
+func LinkController(resourceApisEnabled bool, hcpAllowV2ResourceApis bool, hcpClientFn HCPClientFn) *controller.Controller {
 	return controller.NewController("link", pbhcp.LinkType).
 		WithReconciler(&linkReconciler{
 			resourceApisEnabled:    resourceApisEnabled,
 			hcpAllowV2ResourceApis: hcpAllowV2ResourceApis,
+			hcpClientFn:            hcpClientFn,
 		})
 }
 
 type linkReconciler struct {
 	resourceApisEnabled    bool
 	hcpAllowV2ResourceApis bool
+	hcpClientFn            HCPClientFn
+}
+
+func (r *linkReconciler) linkingFailed(ctx context.Context, rt controller.Runtime, res *pbresource.Resource) error {
+	newStatus := &pbresource.Status{
+		ObservedGeneration: res.Generation,
+		Conditions:         []*pbresource.Condition{ConditionFailed},
+	}
+	if resource.EqualStatus(res.Status[StatusKey], newStatus, false) {
+		return nil
+	}
+	_, err := rt.Client.WriteStatus(ctx, &pbresource.WriteStatusRequest{
+		Id:     res.Id,
+		Key:    StatusKey,
+		Status: newStatus,
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func hcpAccessModeToConsul(mode *gnmmod.HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevel) pbhcp.AccessLevel {
+	if mode == nil {
+		return pbhcp.AccessLevel_ACCESS_LEVEL_UNSPECIFIED
+	}
+
+	switch *mode {
```

Comment:
`mode` is already a pointer. Can we change this and then we can probably get rid of the nil check above?

Reply:
do you mean to change this function so that it looks like:
```golang
func hcpAccessModeToConsul(mode gnmmod.HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevel) pbhcp.AccessLevel {
	switch mode {
	case gnmmod.HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevelCONSULACCESSLEVELUNSPECIFIED:
		return pbhcp.AccessLevel_ACCESS_LEVEL_UNSPECIFIED
	case gnmmod.HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevelCONSULACCESSLEVELGLOBALREADWRITE:
		return pbhcp.AccessLevel_ACCESS_LEVEL_GLOBAL_READ_WRITE
	case gnmmod.HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevelCONSULACCESSLEVELGLOBALREADONLY:
		return pbhcp.AccessLevel_ACCESS_LEVEL_GLOBAL_READ_ONLY
	default:
		return pbhcp.AccessLevel_ACCESS_LEVEL_UNSPECIFIED
	}
}
```
and then just do the nil check before I call this function?

Reply:
Not quite, I think we can just remove the nil check as the default switch case should catch it? `mode` should still be a pointer, though.

Reply:
sorry, I don't think I'm following what you mean. if I just remove the nil check, dereferencing mode in `switch *mode` can panic.
```golang
func hcpAccessModeToConsul(mode *gnmmod.HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevel) pbhcp.AccessLevel {
	//if mode == nil {
	//	return pbhcp.AccessLevel_ACCESS_LEVEL_UNSPECIFIED
	//}

        // panic if mode == nil
	switch *mode {
	case gnmmod.HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevelCONSULACCESSLEVELUNSPECIFIED:
		return pbhcp.AccessLevel_ACCESS_LEVEL_UNSPECIFIED
	case gnmmod.HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevelCONSULACCESSLEVELGLOBALREADWRITE:
		return pbhcp.AccessLevel_ACCESS_LEVEL_GLOBAL_READ_WRITE
	case gnmmod.HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevelCONSULACCESSLEVELGLOBALREADONLY:
		return pbhcp.AccessLevel_ACCESS_LEVEL_GLOBAL_READ_ONLY
	default:
		return pbhcp.AccessLevel_ACCESS_LEVEL_UNSPECIFIED
	}
}
```

Reply:
Ah ok, I just assumed the different cases were already pointers.

That being said, what is the actual underlying type of these values when you say "enum"? Why does `mode` need to be a pointer at all?

Reply:
The different cases are string values that we're pulling in from the `hcp-sdk-go` lib:
```golang
const (

	// HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevelCONSULACCESSLEVELUNSPECIFIED captures enum value "CONSUL_ACCESS_LEVEL_UNSPECIFIED"
	HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevelCONSULACCESSLEVELUNSPECIFIED HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevel = "CONSUL_ACCESS_LEVEL_UNSPECIFIED"

	// HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevelCONSULACCESSLEVELGLOBALREADONLY captures enum value "CONSUL_ACCESS_LEVEL_GLOBAL_READ_ONLY"
	HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevelCONSULACCESSLEVELGLOBALREADONLY HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevel = "CONSUL_ACCESS_LEVEL_GLOBAL_READ_ONLY"

	// HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevelCONSULACCESSLEVELGLOBALREADWRITE captures enum value "CONSUL_ACCESS_LEVEL_GLOBAL_READ_WRITE"
	HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevelCONSULACCESSLEVELGLOBALREADWRITE HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevel = "CONSUL_ACCESS_LEVEL_GLOBAL_READ_WRITE"
)
```

And `mode` is a pointer here because that's how it's returned from the `hcp-sdk-go` code:
```golang
// HashicorpCloudGlobalNetworkManager20220215Cluster hashicorp cloud global network manager 20220215 cluster
//
// swagger:model hashicorp.cloud.global_network_manager_20220215.Cluster
type HashicorpCloudGlobalNetworkManager20220215Cluster struct {
        ...
	// consul_access_level is the level of access CCM has to the Consul cluster
	ConsulAccessLevel *HashicorpCloudGlobalNetworkManager20220215ClusterConsulAccessLevel `json:"consul_access_level,omitempty"`
        ...
}
```

Reply:
>The different cases are string values that we're pulling in from the hcp-sdk-go lib

👍 

>And mode is a pointer here because that's how it's returned from the hcp-sdk-go code

I really don't think these strings should be pointers, but I'm also not going to suggest you rewrite it all. Happy to discuss on Slack, though.

Reply:
we discussed on slack. it seems like that’s how the protobuf -> SDK generation tool we use generates the code/structs. not entirely sure on the reasoning there, but would be a pretty significant change to change that, so leaving as is.